uqmi: added timeouts

dr-ni opened the pull request:
Modems which are using qmi do not reply on the 1st sync but they do on subsequent. Sometimes uqmi is hanging on the first call - even when using an early dummy access to unlock the modem. To guarantee always a proper initialisation, all running or hanging uqmi processes must be stopped.

ptpt52 commented:
what if 2 or more modem connected on the router, instance of uqmi for modem2 may be killed by modem1's

dr-ni commented:
any idea?

ptpt52 commented:
maybe try this is safe:
```
kill -15 $(pgrep -f "uqmi -s -d $device") &>/dev/null
```

dr-ni commented:
ok I'll check this and push it then again

dr-ni commented:
ok, seems to be working nice

dr-ni commented:
I have now pushed an update which is working for me on all of my simcom modems (sim8200ea, sim8202 and sim7600)

hauke commented on package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh:
```diff
@@ -83,9 +83,15 @@ proto_qmi_setup() {
+    timeout 3s for first call to avoid hanging uqmi
+    uqmi -d "$device" --get-pin-status -t 3000 > /dev/null 2>&1
+    # kill pending uqmi
+    kill -15 $(pgrep -f "uqmi -s -d $device") &>/dev/null
```
The indention is wrong.

dr-ni replied:
have it fixed now

hauke commented:
Here @xback suggest to use a timeout of 5 seconds for the WDS command and also some higher timeouts in other places:
https://github.com/openwrt/openwrt/pull/13324#issuecomment-1716424715

Is killing the application still needed after you added the timeouts?

dr-ni commented:
I didn't needed it but to be always on the safe side I recommend this. For examle another uqmi has been initiated from sh

hauke commented:
I am fine with adding the timeouts, but I am not sure if the killing is fine. 
I would prefer if we only add the timeout for now.
Maybe @xback  can also comment on this.

dr-ni commented:
> I am fine with adding the timeouts, but I am not sure if the killing is fine. I would prefer if we only add the timeout for now. Maybe @xback can also comment on this.

can we leave them commented so that one can uncomment them if needed?
can you change this so that I do not have to push again?

dr-ni commented:
the kill commands could also be helpfull when uqmi usb devices are hanging because of unpugging

ptpt52 commented:
if killing is needed, it should change again now:
```
kill -15 $(pgrep -f "uqmi -s -t 1000 -d $device") &>/dev/null
```
or we should change the uqmi commands:
```
uqmi -s -t 1000 -d "$device"  --> uqmi -s -d "$device" -t 1000
```

dr-ni commented:
> if killing is needed, it should change again now:
> 
> ```
> kill -15 $(pgrep -f "uqmi -s -t 1000 -d $device") &>/dev/null
> ```
> 
> or we should change the uqmi commands:
> 
> ```
> uqmi -s -t 1000 -d "$device"  --> uqmi -s -d "$device" -t 1000
> ```

yes you are right

dr-ni commented:
so I'll push it later again

dr-ni commented:
ok, pushed again…

j123b567 commented:
> Is killing the application still needed after you added the timeouts?

Unfortunately yes, see https://github.com/openwrt/openwrt/issues/10472#issuecomment-1400639955 Adding timeouts to `qmi.sh` is a great step forward but not enough, because they are sometimes ignored by `uqmi` binary. There are some combinations of commands and even if you run command `uqmi -t 3000 -d /dev/cdc-wdm0 --some-command` internal command execution will finish after 3 seconds but `uqmi` binary will hang indefinitely.

dr-ni commented:
on my sim8200ea it seems to be working
```
root@OpenWrt:~# uqmi -t 3000 -d /dev/cdc-wdm0 --get-data-status
"disconnected"
root@OpenWrt:~# uqmi -t 3000 -d /dev/cdc-wdm0 --get-signal-info
Request timed out
"Failed to connect to service"
root@OpenWrt:~# uqmi -t 3000 -d /dev/cdc-wdm0 --get-signal-info
{
	"type": "lte",
	"rssi": -65,
	"rsrq": -15,
	"rsrp": -97,
	"snr": -1.400000
}
```
which commands are hanging?

j123b567 commented:
It is not so easy to reproduce. Multiple (mostly unknown) preconditions must be met. If you put the modem in the "right" state and send the "right" command at the "right" time you can trigger that bug. Timing depends on the SIM card, operator, and signal strength. You can try to reproduce this by some sort of stress test.

We have a fleet with tens of devices with SIM7500e and it does not happen every day but it was annoying enough that we put the command `timeout` before calling `uqmi` which solved the problem for us. Some of our devices were more susceptible to this bug than others.

dr-ni commented:
@hauke

how shall we proceed?

j123b567 commented:
Sorry for the interruption. IMHO the correct way would be to pretend here that `uqmi` can handle timeouts correctly (no killing). And fix `uqmi` itself to do so (or add workarounds there).

xback commented:
> Sorry for the interruption. IMHO the correct way would be to pretend here that `uqmi` can handle timeouts correctly (no killing). And fix `uqmi` itself to do so (or add workarounds there).

I agree on this one.
While all these changes help as a bandage to avoid the issue, they are by no means an actual fix
for the issues I [described earlier](https://github.com/openwrt/openwrt/pull/13324#issuecomment-1716424715).

So basically, this change is fine for me, as timeouts really need to be set on a per-command base for fast response if something goes wrong inside the modem itself, but additional fixing **is** needed in side uqmi itself. (or even libubox, as functions from this one are used for the actual reading/writing)

I really really would love to post a new tool based on my internal code .. which would add auto-detection, automatic multi-modem support, mixed QMI/AT (for GNSS support), etc, .. but I need to request that internally at work.

dr-ni commented:
ok, @xback a tool you discribed would be very nice :-)
So from my side I would be saying we can now merge…

Leo-PL commented:
I don't want to be picky, but the commented out calls to `kill` now present in https://github.com/openwrt/openwrt/commit/728dfd176ccb28fc15f1b27cbb592086956033a9 should be removed.
I don't like how this change, and the previous one is handled, trying to rush the solution in, instead of engineering one.

dr-ni commented:
> I don't want to be picky, but the commented out calls to `kill` now present in [728dfd1](https://github.com/openwrt/openwrt/commit/728dfd176ccb28fc15f1b27cbb592086956033a9) should be removed. I don't like how this change, and the previous one is handled, trying to rush the solution in, instead of engineering one.

The reason for this can be simply be  explained:

I was using my first qmi-based modem since january this year and noticed the issues with uqmi.
Searching the web for solutions I stombled over this:

https://forum.openwrt.org/t/solved-lte-qmi-troubles/28379/24 (from Jan 2019).

After about 4 years now without any improvements I do not expect a clean engineering solution within the next time {the first discussion was in Jan 2019).

But no problem, I can remove the comments.

dr-ni commented:
Is this request not wanted anymore?
It would be nice if anybody cozld review or close it

Commit:
uqmi: added timeouts in qmi.sh

Modems which are using qmi do not reply on the 1st sync but they do
on subsequent. Sometimes uqmi is hanging - even when using an early
dummy access to unlock the modem. To always guarantee a proper
initialisation, running or hanging uqmi processes must be stopped
before. All uqmi calls have now a timeout option -t to avoid hanging.

Signed-off-by: Uwe Niethammer <uwe@dr-niethammer.de>

dr-ni commented:
now again with latest upstream changes...

dr-ni commented:
Oh no :-( the upstream version is not working anymore with my sim8200ea
since 79912a953111dee20a6f8abdf31a4fc25a9b160e I get in luci: `Error: Unknown error (SIM_ILLEGAL_STATE)`
If I switch back to my previous version it's working again but now only very slow speeds
What is going on?

dr-ni commented:
Ok very strange:
I needed a full factory reset of my modem:

```
root@OpenWrt:/lib/netifd/proto# minicom -D /dev/ttyUSB2


Welcome to minicom 2.8

OPTIONS: 
Port /dev/ttyUSB2, 03:42:32

Press CTRL-A Z for help on special keys
                                                        
at+cusbcfg=usbid,1e0e,9001                              
OK                                                      
AT                                                      
OK     
```
Now the speed is back (from 12 to 200 Mbit)
The upstream version is changing my modem settings?

hauke commented:
@blocktrron  Could you please have a look at 696c9f723d3b2fe1577da849bab97a93b9ad5d68.

dr-ni commented:
> @dr-ni Is the change from this pull request working for your only if you revert [696c9f7](https://github.com/openwrt/openwrt/commit/696c9f723d3b2fe1577da849bab97a93b9ad5d68)?

yes

dr-ni commented:
update: the release 23.05.2 has now the new uqmi option implementated.
However, the qmi.sh without the timeouts is still hanging.
I'm currently testing this PR qmi.sh …

dr-ni commented:
on first start I'm watching this:

- my 5g interface is starting and running for 10 seconds
- in luci I'm getting the message `device not present
- then my 5g interface is stopped again
- after some time I have this message `Error: Unknown error (SIM_ILLEGAL_STATE)` again

blocktrron commented:
@hauke I'll have a look.

@dr-ni Can you provide me with an output of `--uim-get-sim-state` on your modem? I suspect your device does not implement the card-application-state output, thus always triggering the SIM recovery.

dr-ni commented:
Before running 5g device:
```
root@OpenWrt:~# uqmi -t 7000 -d /dev/cdc-wdm0 --uim-get-sim-state
Request timed out
"Failed to connect to service"
```
after restart in luci:
```
root@OpenWrt:~# pgrep qmi 
8543
8629
root@OpenWrt:~# pgrep uqmi 
root@OpenWrt:~# pgrep qmi 
8543
8835
8844
root@OpenWrt:~# pgrep qmi
8543
root@OpenWrt:~# uqmi -t 7000 -d /dev/cdc-wdm0 --uim-get-sim-state
{
	"card_application_state": "ready",
	"pin1_status": "disabled",
	"pin1_verify_tries": 3,
	"pin1_unlock_tries": 10,
	"pin2_status": "not_verified",
	"pin2_verify_tries": 3,
	"pin2_unlock_tries": 10
}
```

blocktrron commented:
@dr-ni you say "release 23.05.2 has now the new uqmi option implemented" but the illegal sim-state recovery is only implemented in master. Yan you clarify which version you are actually observing this behavior on?

The messages in LuCI are not really helpful, a syslog output would make it better understandable what exactly is working on.

> I think all these merged PRs should be reverted since they are not tested well

I'm begging to differ on this one. We've tested it on UC20, EC25 EP06 and RG510 modems and it worked on all those models.

Dual-SIM modems usually switch sim-cards based on a mux triggered by a GPIO from the host, I've yet to encounter a modem which actually supports on-board sim switching.

Leo-PL commented:
> Dual-SIM modems usually switch sim-cards based on a mux triggered by a GPIO from the host, I've yet to encounter a modem which actually supports on-board sim switching.

Sierra EM7455 does, albeit it needs some AT commands to allow that in the first place. Available cheaply at local auction sites, in Poland they come for an equivalent of around 5EUR used ;-]

dr-ni commented:
@blocktrron I'm using master (OpenWrt 23.05.2, r23630-842932a63d) and this qmi.sh https://raw.githubusercontent.com/openwrt/openwrt/8c24ae4681813e095415583c62a17577ad30cbf6/package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh

If I will have time I  can send you also a logfile and I can perform some tests for dual sim if you give me some advice…

ptpt52 commented:
I think it is safe to use `-t 3000` instead of `-t 1000`

dr-ni commented:
I'm afraid - but the progress here is stocking
I could change this again but this PR was also stocking before my last change.
So it was then outdated and with conflicts - at the end I have always the effort for nothing…

So hey guys PLEASE tell me how to proceed!

dr-ni commented:
is there any interest in continuing this?

hauke commented:
I am still confused in which version you saw problems with the SIM and which version your tested. 
Did you build your branch dr-ni:master and then saw the problems with the SIM card? 
Did you use OpenWrt 23.05 stable release and then modified the `qmi.sh` script with your changes?

Leo-PL commented:
I can only add, that with the version merged to current master, I still observed the issues with my ZTE modems - but downstream patch carried by @obsy fixed them - which is slightly more elaborate. I need to get back to this and implement @blocktrron's suggestion properly.

hauke commented:
I am not using qmi, as far as I understood this pull request is fine, but there are some other problems. 

@Leo-PL Which patch from @obsy  are you using?

Leo-PL commented:
@hauke it is stashed in here, with others:
https://dl.eko.one.pl/openwrt-23.05/patches/r23695-80ef582dee.patch
```
diff --git a/package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh b/package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh
index eab7f1ea13..f0665aa375 100755
--- a/package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh
+++ b/package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh
@@ -85,7 +85,7 @@
 	local uninitialized_timeout=0
 	# timeout 3s for first call to avoid hanging uqmi
 	uqmi -d "$device" --get-pin-status -t 3000 > /dev/null 2>&1
-	while uqmi -s -d "$device" --get-pin-status | grep '"UIM uninitialized"' > /dev/null; do
+	while uqmi -t 3000 -s -d "$device" --get-pin-status | grep '"UIM uninitialized"' > /dev/null; do
 		[ -e "$device" ] || return 1
 		if [ "$uninitialized_timeout" -lt "$timeout" -o "$timeout" = "0" ]; then
 			let uninitialized_timeout++
@@ -98,8 +98,8 @@
 		fi
 	done
 
-	if uqmi -s -d "$device" --uim-get-sim-state | grep -q '"Not supported"\|"Invalid QMI command"' &&
-	   uqmi -s -d "$device" --get-pin-status | grep -q '"Not supported"\|"Invalid QMI command"' ; then
+	if uqmi -t 3000 -s -d "$device" --uim-get-sim-state | grep -q '"Not supported"\|"Invalid QMI command"' &&
+	   uqmi -t 3000 -s -d "$device" --get-pin-status | grep -q '"Not supported"\|"Invalid QMI command"' ; then
 		[ -n "$pincode" ] && {
 			uqmi -s -d "$device" --verify-pin1 "$pincode" > /dev/null || uqmi -s -d "$device" --uim-verify-pin1 "$pincode" > /dev/null || {
 				echo "Unable to verify PIN"
@@ -109,10 +109,10 @@
 			}
 		}
 	else
-		json_load "$(uqmi -s -d "$device" --get-pin-status)"
+		json_load "$(uqmi -t 3000 -s -d "$device" --get-pin-status)"
 		json_get_var pin1_status pin1_status
 		if [ -z "$pin1_status" ]; then
-			json_load "$(uqmi -s -d "$device" --uim-get-sim-state)"
+			json_load "$(uqmi -t 3000 -s -d "$device" --uim-get-sim-state)"
 			json_get_var pin1_status pin1_status
 		fi
 		json_get_var pin1_verify_tries pin1_verify_tries
```
Basically, it's simplification of @blocktrron's idea, but it's enough to get my ZTE modems running well.

dr-ni commented:
@hauke
I was using the stable master (OpenWrt 23.05.2, r23630-842932a63d) and replaced qmi.sh with this one here

hauke commented:
@Leo-PL Your change looks similar to this one, you are just adding a timeout to some functions and you are using 3 seconds. Are you find with the change from this pull request adding 1 second and adding it to all calls or does your change have any advantages? did you see that it needs more than 1 second in some cases?

Leo-PL commented:
According to @xback, certain commands can take longer, see here: https://github.com/openwrt/openwrt/pull/13324#issuecomment-1716424715 - I believe the timeouts should be increased.

dr-ni commented:
I think this is already in: https://github.com/openwrt/openwrt/pull/13781/files

the question is if all '1000' timeouts should be '3000'

hauke commented:
Thank you for the PR, I merged it into main in 6b0db8592a3e4342c32111491948f32d5bc0087f.

yogo1212 commented:
fyi, the timeout option (sadly) doesn't resolve deadlocks with multiple parties using a modem (in fact, i've found it to increase their likelihood).

lynxis is working on a solution that manages the modem, synchronizes access to the device, and exposes an ubus interface.

yogo1212 commented:
another note:

i was working with uqmi two years ago and also went down this route.

uqmi can (and will) still hang - which is why i came up with this to put at the beginning of the file:

```shell
proto_qmi_inject_uqmi() {
  uqmi() {
    local t="${TIMEOUT-10}"

    timeout -k "$(( t + 2 ))" "$(( t + 1 ))" /sbin/uqmi -t "$(( t * 1000 ))" "$@"
  }
}
```

that (and other patches) are in [here](https://github.com/openwrt/openwrt/pull/10906).
